tiledb/sm/filesystem/s3.h

```diff
@@ -271,6 +336,199 @@ struct S3Parameters {
+      bool use_virtual_addressing)
+      : Aws::S3::S3Client(
+            creds, client_config, sign_payloads, use_virtual_addressing)
+      , params_(s3_params) {
+  }
+
+  void BuildHttpRequest(
+      const Aws::AmazonWebServiceRequest& request,
+      const std::shared_ptr<Aws::Http::HttpRequest>& httpRequest)
+      const override {
+    S3Client::BuildHttpRequest(request, httpRequest);
+
+    // Set header from S3Parameters custom headers
+    for (auto& [key, val] : params_.custom_headers_) {
+      httpRequest->SetHeaderValue(key, val);
+    }
+  }
+
+  inline bool requester_pays() const {
+    return params_.requester_pays_;
+  }
+
+ protected:
+  /**
+   * A reference to the S3 configuration parameters, which stores the header.
+   *
+   * @note Until the removal of init_client(), this must be const-qualified.
+   */
+  const S3Parameters& params_;
+};
+
+/**
+ * S3Scanner wraps the AWS ListObjectsV2 request and provides an iterator for
+ * results. If we reach the end of the current batch of results and results are
+ * truncated, we fetch the next batch of results from S3.
+ *
+ * @tparam F The FilePredicate type used to filter object results.
+ * @tparam D The DirectoryPredicate type used to prune prefix results.
+ */
+template <FilePredicate F, DirectoryPredicate D = DirectoryFilter>
+class S3Scanner : public LsScanner<F, D> {
+ public:
+  /** Declare LsScanIterator as a friend class for access to call next(). */
+  template <class T, class U>
+  friend class LsScanIterator;
+  using Iterator = LsScanIterator<S3Scanner<F, D>, Aws::S3::Model::Object>;
+
+  /** Constructor. */
+  S3Scanner(
+      const std::shared_ptr<TileDBS3Client>& client,
+      const URI& prefix,
+      F file_filter,
+      D dir_filter = accept_all_dirs,
+      bool recursive = false,
+      int max_keys = 1000);
+
+  /**
+   * Returns true if there are more results to fetch from S3.
+   */
+  [[nodiscard]] inline bool more_to_fetch() const {
+    return list_objects_outcome_.GetResult().GetIsTruncated();
+  }
+
+  /**
+   * Constructs an iterator to the beginning of the results for this scan.
+   *
+   * @return Iterator to the beginning of the results for this scan.
+   */
+  Iterator iterator() {
+    return Iterator(this);
+  }
+
+ private:
+  /**
+   * Advance to the next object accepted by the filters for this scan.
+   *
+   * @param ptr Reference to the current data pointer.
+   * @sa LsScanIterator::operator++()
+   */
+  void next(typename Iterator::pointer& ptr);
+
+  /**
+   * Advance the iterator to the next object in the current batch of results.
+   * If the iterator is at the end of the current batch, this will fetch the
+   * next batch of results from S3. This does not check if the results are
+   * accepted by the filters for this scan.
+   *
+   * @param ptr Reference to the current data pointer.
+   */
+  void advance(typename Iterator::pointer& ptr) {
+    ++ptr;
+    if (ptr == end_) {
```

Comment:
This way of detecting end-of-container won't work, because the address is outside the bounds of the container.

This isn't hypothetical. We've had exactly this class of defect before. It happened when an allocation was right at the end of the memory block. A computed address of this form was an illegal address and would fail by segfault in certain circumstances.